From ddeb18fa217b94959f66a4924decce69e0e97d4e Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Thu, 10 Oct 2024 10:59:24 +0200
Subject: [PATCH 1285/1695] drivers: stm32_remoteproc: Don't use SRAMs used by
 pager

Forbid stm32_remoteproc driver to use SRAMx that are used by OP-TEE pager.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 3ecd94fdef5a0de38bf316bb7d0ac7f739a516cf drivers: stm32_remoteproc: Don't use SRAMs used by pager]
Change-Id: Iee6f81a8c45e61aec214f74a8a98969dc186f4ac
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/410434
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/remoteproc/stm32_remoteproc.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/core/drivers/remoteproc/stm32_remoteproc.c b/core/drivers/remoteproc/stm32_remoteproc.c
index 420e4800a..8a339a3a3 100644
--- a/core/drivers/remoteproc/stm32_remoteproc.c
+++ b/core/drivers/remoteproc/stm32_remoteproc.c
@@ -20,6 +20,7 @@
 #if defined(CFG_STM32MP25) || defined(CFG_STM32MP23) || defined(CFG_STM32MP21)
 #include <stm32_sysconf.h>
 #endif
+#include <stm32_util.h>
 
 #define TIMEOUT_US_1MS	U(1000)
 
@@ -613,6 +614,17 @@ static TEE_Result stm32_rproc_parse_mems(struct stm32_rproc_instance *rproc,
 			res = TEE_ERROR_GENERIC;
 			return res;
 		}
+
+#ifdef CFG_STM32MP15
+		if (stm32mp1_ram_intersect_pager_ram(regions[i].addr,
+						     regions[i].size)) {
+			EMSG("Region %#"PRIxPA"..%#"PRIxPA" intersects pager secure memory",
+			     regions[i].addr,
+			     regions[i].addr + regions[i].size);
+			return TEE_ERROR_GENERIC;
+		}
+#endif
+
 		res = firewall_dt_get_alternate_conf(fdt, pnode,
 						     FIREWALL_CONF_DEFAULT,
 						     &regions[i].default_conf);
-- 
2.47.2

